netty/src/main/java/io/grpc/netty/NettyServerBuilder.java

```diff
@@ -492,6 +493,41 @@ public NettyServerBuilder maxHeaderListSize(int maxHeaderListSize) {
   public NettyServerBuilder maxInboundMetadataSize(int bytes) {
     checkArgument(bytes > 0, "maxInboundMetadataSize must be positive: %s", bytes);
     this.maxHeaderListSize = bytes;
+    // Clear the soft limit setting, by setting soft limit to maxInboundMetadataSize. The
+    // maxInboundMetadataSize will take precedence be applied before soft limit check.
+    this.softLimitHeaderListSize = bytes;
+    return this;
+  }
+
+  /**
+   * Sets the size of metadata that advised to not exceed. When a metadata with size larger than the
+   * soft limit is encountered there will be a probability the RPC will fail. The chance of failing
+   * increases as the metadata size approaches the hard limit. {@code Integer.MAX_VALUE} disables
+   * the enforcement. The default is implementation-dependent, but is not generally less than 8 KiB
+   * and may be unlimited.
+   *
+   * <p>This is cumulative size of the metadata. The precise calculation is
+   * implementation-dependent, but implementations are encouraged to follow the calculation used
+   * for
+   * <a href="http://httpwg.org/specs/rfc7540.html#rfc.section.6.5.2">HTTP/2's
+   * SETTINGS_MAX_HEADER_LIST_SIZE</a>. It sums the bytes from each entry's key and value, plus 32
+   * bytes of overhead per entry.
+   *
+   * <P>Note: gRPC has approximately 600B internal Metadata.
```

Comment:
We either need to drop this or have it on the other overloads of the method. Honestly, it varies quite a bit and is I'm worried people will be like "gRPC used more than 600 bytes, which is more than it said." So I'd favor leaving it out. When people call this they're likely to set it to 100s of KiB or a MiB anyway.

Reply:
Done.